### ISSUE TITLE
[FoxyTab] Add option for "close duplicate tabs" to be container aware

### ISSUE THREAD

**Issue by WhonderWy** [erosman/support#289]
I use containers like many others to manage my online identities and assist with tracking mitigation.
I recently realised that the "Close Duplicate Tabs" action closes duplicates regardless of the containers they're in. A bit of a shame but I can see it being a bigger problem the more one uses containers as it would mean finding each site and reopening them in the relevant container.

I propose adding an option so that users can choose whether "Close Duplicate Tabs" closes *every* duplicate tab or closes duplicates only if they're in the same container, or not in a container at all.

In more exact terms [pseudocode since I don't know any of the actual variable names used or the language]:
* Option:
```
if container_aware_option is toggled on:
    set container_aware_option as true
```
* Actual function:
```
// Somewhere around the duplicate checking logic which probably involves checking for same url.
// Basically appending a check that the container is the same
if container_aware_option is true:
    if ((this_url equals that_url) and (this_url_container equals that_url_container)):
        // close the duplicate tab as usual.
```

Since there's no issue template, I just wrote this how I felt. Sorry if it's a difficult read.
Also, I haven't thought much about any bugs or issues arising out of this case, so I do apologise for that.

TLDR: As a container user, I want the "Close Duplicate Tabs" to be container aware and not close duplicates if they're in different containers.

**Comment by erosman** [comment]
It is possible to add such feature but there are too many variations.

- Close all duplicates
- Close all non-container duplicates
- Close all container duplicates
- Close all container A duplicates
- Close all container B duplicates
- Close all container C duplicates
-  .....

It would require many menu items to cover all the possibilities or else it runs into the situation where users would ask why does it do this and not that.

**Comment by WhonderWy** [comment]
Isn't those variations sort of different and/or not necessary [at least for my proposed use case (perhaps if people are interested, they could vote for that instead)]? Multi-account containers currently generate a list if you click on the inline `>` from which you can delete tabs from easily. I personally don't have as many tabs in containers as outside of them, so my current proposal would be as outlined above; that is to say that a [default: off] switch toggles from the current behaviour to an on state which ignores container duplicates.

On the other hand, I'm all for new features so while it doesn't quite match my proposal in scope or intention, I feel like commenting on the variations proposed. If my proposal is implemented, at the very least, the close all duplicates function should change to be more specific and then the current behaviour could be implemented by wrapping the new function and looping through all the container tabs as well. Similarly, choosing any specific container to delete duplicates of would entail passing the relevant container information as well.
As for the menu, would changing it to a self-populating list based on the user's currently active or created containers make a difference? So for example, the plugin would get a list of containers that it appends [in any order you like]: the current [close all duplicates] behaviour, only container duplicates and non-container duplicates options to. This list would be displayed either in the options page or as a popup or even within the menu by changing the current "close duplicate tabs" item into a "duplicate tabs" [which could make #199 more viable by not cluttering existing menus maybe? (so duplicate tab in a private window)]. I'm envisioning something like:
```
- Duplicate Tab
- Close Duplicate Tabs
- -------------------------
- Close Tabs to the Left
```
could be turned into something like:
```
- Duplicate Tab Actions
- - Duplicate (this) Tab
- - Duplicate/Copy (this) Tab to Private Window
// Add Left/Right options here
- - Close Duplicate Tabs
- - - Close all duplicates
- - - Close all non-container duplicates
- - - Close all container duplicates
- - - Close all container A duplicates
- - - Close all container B duplicates
- - - Close all container C duplicates
- - - ... etc.
```
Or if it gets too busy:
```
- Duplicate Tab Actions
- - Duplicate (this) Tab
- - Duplicate/Copy (this) Tab to Private Window
// Add Left/Right options here
- - Close Duplicate Tabs
- - - Close all duplicates
- - - Close all non-container duplicates
- - - Close all container duplicates
- - - Custom // which opens a menu or something where you can choose which container duplicates to close [maybe even more than one!]
```
It would require changing the implementation of the current `Close Duplicate Tabs`-related function, but it would be something I'd be happy to get/see in the addon or even as part of Firefox itself.

Sorry, I'm talking a lot based on my imagination and half-baked programming knowledge since I don't know what the code looks like and probably wouldn't be able to alter it myself since I swore myself off JavaScript. I want to see it implemented, so I'm putting ideas up, but I'm fully aware it's your time that's spent on it.

**Comment by erosman** [comment]
Let's see if there is a popular demand for the feature.

**Comment by erosman** [comment]
I am working to prepare v2.16 for release.

- Added Container support to actions

When `Ctrl` key is pressed while clicking on FoxyTab context-menus, actions would become **limited** to the container/contextual-identity of the clicked tab (or active tab in case of toolbar icon context-menu), for all applicable actions.

Therefore, holding `Ctrl` key & clicking `Close Duplicate Tabs` on container A tab, will close duplicates in container A tabs only.
